### PR TITLE
[cockroachdb] Add support for single node, document minumum number of supported nodes

### PR DESCRIPTION
**Pull request by jonaseck2**
Signed-off-by: Jonas Eckerström <jonaseck@gmail.com>

#### What this PR does / why we need it:

When specifying number of replicas less than 3, health checks won't pass since cluster members fail to join the specified cluster members since the number of nodes to join is hard coded to 3.

If number of replicas is 1, --join is omitted from the start command entirely
Added JoinReplicas specifying the number of nodes to join, as long as it's less than or equal to number of replicas, we can now start a cluster of any size.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

File: stable/cockroachdb/README.md [hunk 1]
```diff
@@ -101,6 +101,7 @@ The following table lists the configurable parameters of the CockroachDB chart a
 | `Secure.ServiceAccount.Create` | Whether to create a new RBAC service account     | `true`                                    |
 | `Secure.ServiceAccount.Name`   | Name of RBAC service account to use              | `""`                                      |
 | `JoinExisting`                 | List of already-existing cockroach instances     | `[]`                                      |
+| `JoinReplicas`                 | number of replicas to join, not used if `JoinExisting` is specified  | `3`                   |
```
Comment (a-robinson) [inline stable/cockroachdb/README.md]: Do you actually want a custom number of `JoinReplicas`? Or is the change to only specify join flags if the number of replicas is greater than 1 good enough? My intuition would be that this new parameter is unlikely to be used, and that your new conditional is the right fix.

File: stable/cockroachdb/templates/cockroachdb-statefulset.yaml [hunk 2]
```diff
@@ -290,12 +290,13 @@ spec:
         - name: certs
           mountPath: /cockroach/cockroach-certs
 {{- end }}
+{{- $root := . }}
```
Comment (a-robinson) [inline stable/cockroachdb/templates/cockroachdb-statefulset.yaml]: Why is this needed now?

**Review by a-robinson — CHANGES_REQUESTED** [review]

**Comment by jonaseck2** [comment]
@a-robinson Yeah it's a bit overkill, I actually just need support for single node development cluster support by removing the join flag for 1 replica.

The reason for finally adding a JoinReplicas value was actually because we'd end up with a cluster that 1 works (due to no join), 2 would never be healthy and 3 or more would work. So do we move forward and implement a math.max(replicas, 3) to join with or support the case for 1 host and no join?

**Comment by a-robinson** [comment]
Actually, I'm confused. Does the chart really not work with a cluster size of 1 replica? The fact that extra nodes are specified in the join flags shouldn't matter; the first node in the cluster should be properly initialized by [the post-install init job](https://github.com/helm/charts/blob/master/stable/cockroachdb/templates/cluster-init.yaml).

Apologies if I'm misunderstanding -- my head is somewhat far removed from this right now.

**Comment by jonaseck2** [comment]
@a-robinson yes, when setting a lower number of replicas to join than 3, the cluster never joins up and the pods never become healthty

I've changed my contribution to only add a --join flag if you have more replicas than 1. You'd still end up with an unhealthy cluster if you set 2 replicas, so i've documented what's supported.

File: stable/cockroachdb/README.md [hunk 3]
```diff
@@ -71,7 +71,7 @@ The following table lists the configurable parameters of the CockroachDB chart a
 | `Image`                        | Container image name                             | `cockroachdb/cockroach`                   |
 | `ImageTag`                     | Container image tag                              | `v2.1.5`                                  |
 | `ImagePullPolicy`              | Container pull policy                            | `Always`                                  |
-| `Replicas`                     | k8s statefulset replicas                         | `3`                                       |
+| `Replicas`                     | k8s statefulset replicas, `1` or at least`3`     | `3`                                       |
```
Comment (a-robinson) [inline stable/cockroachdb/README.md:74]: Looks like you're missing a space after the word "least"

File: stable/cockroachdb/templates/cockroachdb-statefulset.yaml [hunk 4]
```diff
@@ -300,7 +300,7 @@ spec:
...
+          - "exec /cockroach/cockroach start --logtostderr {{ if .Values.Secure.Enabled }}--certs-dir /cockroach/cockroach-certs{{ else }}--insecure{{ end }} --advertise-host $(hostname).${STATEFULSET_FQDN} --http-host 0.0.0.0 --http-port {{ .Values.InternalHttpPort }} --port {{ .Values.InternalGrpcPort }} --cache {{ .Values.CacheSize }} --max-sql-memory {{ .Values.MaxSQLMemory }} {{ if .Values.Locality }}--locality={{.Values.Locality }}{{ end }} {{ if or (gt (int .Values.Replicas) 1) .Values.JoinExisting }} --join {{ if .Values.JoinExisting }}{{ join "," .Values.JoinExisting }}{{ else }}${STATEFULSET_NAME}-0.${STATEFULSET_FQDN}:{{ .Values.InternalGrpcPort }},${STATEFULSET_NAME}-1.${STATEFULSET_FQDN}:{{ .Values.InternalGrpcPort }},${STATEFULSET_NAME}-2.${STATEFULSET_FQDN}:{{ .Values.InternalGrpcPort }}{{end}}{{ end }}{{ range .Values.ExtraArgs }} {{ . }}{{ end }}"
```
Comment (a-robinson) [inline stable/cockroachdb/templates/cockroachdb-statefulset.yaml:303]: I'm having a hell of a time parsing this. Could you perhaps split it onto separate lines? I imagine some backslashes or other formatting characters will be needed, but it started making sense when I formatted it something like this

```
{{ if or (gt (int .Values.Replicas) 1) .Values.JoinExisting }}
    --join 
    {{ if .Values.JoinExisting }}
        {{ join "," .Values.JoinExisting }}
    {{ else }}
        ${STATEFULSET_NAME}-0.${STATEFULSET_FQDN}:{{ .Values.InternalGrpcPort }},${STATEFULSET_NAME}-1.${STATEFULSET_FQDN}:{{ .Values.InternalGrpcPort }},${STATEFULSET_NAME}-2.${STATEFULSET_FQDN}:{{ .Values.InternalGrpcPort }}
    {{end}}
{{ end }}
```

**Review by a-robinson — COMMENTED** [review]
> @a-robinson yes, when setting a lower number of replicas than 3, the cluster never joins up and the pods never become healthty

Do you mean when also using `JoinExisting` or something like that? Because I just installed the latest versions on Minikube and Helm and tried it out myself, and running `helm install stable/cockroachdb --set Replicas=1` successfully created a perfectly functional cluster. I was able to use the command in the `helm install` output to open a shell up against the cluster and everything.

Can you be more specific about what isn't working?

**Comment by a-robinson** [comment]
And I'm sorry for the delay on this, I don't work on this stuff as part of my day job anymore and I've been quite busy lately.

**Pull request closed without merging**